File: sorting/c/heap.c
#ifdef _WIN32
#include "array.c"
#elif __APPLE__
#include "array.h"
#elif __linux__
#include "array.h"
#endif

#include <stdio.h>

void heapify(Arr_T A, int N, int i) {
  int largest = i;
  int left = (2 * i) + 1;
  int right = (2 * i) + 2;
  if (left < N && A.arr[left] > A.arr[largest]) {
    largest = left;
  }
  if (right < N && A.arr[right] > A.arr[largest]) {
    largest = right;
  }
  if (largest != i) {
    swap(i, largest, A);
    heapify(A, N, largest);
  }
}

void heapSort(Arr_T A, int N) {
  for (int i = (N / 2 - 1); i >= 0; i--) {
    heapify(A, N, i);
  }
  for (int j = (N - 1); j >= 0; j--) {
    swap(0, j, A);
    heapify(A, j, 0);
  }
}

void sort(Arr_T A) { heapSort(A, A.size); }

#ifndef ASLIB
int main(int argc, char *argv[]) {
  int arr_size;
  if (argc > 1) {
    if (sscanf(argv[1], "%i", &arr_size) != 1) {
      fprintf(stderr, "[ERR] - Not an integer!");
      return 1;
    }
  } else {
    arr_size = 100;
  }
  Arr_T M = make_Arr(arr_size);
  populate_Arr(M);
  printf("Array before sorting: \n");
  print_Arr(M);
  heapSort(M, M.size);
  printf("Array after sorting: \n");
  print_Arr(M);
  return 0;
}
#endif

File: sorting/python/heap.py
import array
import time


def heapify(A, N: int, i: int):
    largest = i
    left = (2 * i) + 1
    right = (2 * i) + 2
    if left < N and A.arr[left] > A.arr[largest]:
        largest = left
    if right < N and A.arr[right] > A.arr[largest]:
        largest = right
    if largest != i:
        A.swap(i, largest)
        heapify(A, N, largest)


def heap_sort(A, N: int):
    for i in range(int(N / 2 - 1), -1, -1):
        heapify(A, N, i)
    for j in range(int(N - 1), -1, -1):
        A.swap(0, j)
        heapify(A, j, 0)


if __name__ == "__main__":
    array = array.Arr_T()
    array.make_Arr(100)
    array.populate_Arr()
    print("Array before sorting: ")
    array.print_Arr()
    start = time.time()
    heap_sort(array, array.size)
    end = time.time()
    print("\nArray after sorting: ")
    array.print_Arr()
    print("\nThe execution time for the heap sort algorithm is: " + str(float(end-start)))
